database/dbindex.go

```diff
@@ -0,0 +1,184 @@
+package database
+
+import (
+	"github.com/blang/semver"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/globalsign/mgo"
+	"github.com/globalsign/mgo/bson"
+)
+
+type (
+	// RITADatabaseIndex tracks registered RITA databases
+	RITADatabaseIndex struct {
+		dbHandle            *mgo.Session
+		metaDatabaseName    string
+		indexCollectionName string
+		log                 *log.Logger
+	}
+)
+
+// NewRITADatabaseIndex creates a collection in the MetaDatabase
+// to track RITA managed databases. Most importantly, the database index
+// contains information on whether databases have been imported,
+// and/ or analyzed.
+func NewRITADatabaseIndex(dbHandle *mgo.Session,
+	metaDatabaseName string, indexCollectionName string,
+	logger *log.Logger) (RITADatabaseIndex, error) {
+
+	// ensure the index collection is indexed by name
+	nameIndex := mgo.Index{
+		Key:        []string{"name"},
+		Unique:     true,
+		DropDups:   true,
+		Background: true, // This is a hold over. I don't think we want this. (LL)
+		Name:       "nameindex",
+	}
+
+	//creating an index is idempotent so it won't error if the index already
+	//exists. Also collections are created implicitly, so we can create
+	//the collection and ensure the index exists in one fell swoop.
+	err := dbHandle.DB(metaDatabaseName).C(indexCollectionName).EnsureIndex(nameIndex)
+	if err != nil {
+		logger.WithFields(log.Fields{
+			"metadatabase":     metaDatabaseName,
+			"index_collection": indexCollectionName,
+			"error":            err.Error(),
+		}).Error("Failed to create index for the database index")
+		return RITADatabaseIndex{}, err
+	}
+
+	index := RITADatabaseIndex{
+		dbHandle:            dbHandle.Copy(),
+		metaDatabaseName:    metaDatabaseName,
+		indexCollectionName: indexCollectionName,
+		log:                 logger,
+	}
+	return index, nil
+}
+
+func (r *RITADatabaseIndex) openIndexCollection() *mgo.Collection {
```

Comment:
As a comment (I wouldn't say it's needed now) we should move away from single character variable names. It makes lines longer but improves readability considerably (even for one line functions like this

Reply:
That makes sense.  I'm not sure it would add much benefit for a tiny function like this but it's probably easiest to stick with one guideline instead of a guideline with a bunch of exceptions.

Since it sounds like single character variables are fairly common elsewhere in RITA, I agree this doesn't need to be the time to address it.  Though we could do with a style guide that covers these types of things (or better yet some custom linting rules).  A refactor would be much easier to do at one point in time after we've formalized something like that.

Reply:
As a note for future discussion, the [Effective Go](https://golang.org/doc/effective_go.html) guide and the [Go standard library](https://github.com/golang/go/blob/master/src/compress/zlib/reader.go) use single letter variable names for the method receiver. However, I agree that single letter variable names can be confusing. Especially if you use common indexing variables `i`, `j`, or `n`.